File: Irrgarten_Java/src/irrgarten/LabyrinthCharacter.java
package irrgarten;

/**
 * Clase que representa los distintos personajes del juego que se encuentran
 * en el laberinto.
 * 
 * @author Arturo Olivares Martos
 * @author Joaquín Avilés de la Fuente
 */

abstract class LabyrinthCharacter {
    /**
     * Nombre del personaje del laberinto.
     */
    private String name;
    /**
     * Inteligencia del personaje del laberinto.
     */
    private float intelligence;
    /**
     * Fuerza del personaje del laberinto.
     */
    private float strength;
    /**
     * Salud del personaje del laberinto.
     */
    private float health;
    /**
     * Fila en la que se encuentra el personaje del laberinto.
     */
    private int row;
    /**
     * Columna en la que se encuentra el personaje del laberinto.
     */
    private int col;
    
    /** 
     * Posición inválida para inicializar row y col en el constructor.
     */
    private static final int INVALID_POS=-1;
    
    /**
     * Constructor de la clase abstracta LabytinthCharacter
     * @param name Valor inicial para el nombre
     * @param intelligence Valor inicial para la inteligencia
     * @param strength Valor inicial para la fuerza
     * @param health Valor inicial para la salud
     */
    public LabyrinthCharacter (String name, float intelligence, float strength, float health){
        this.name=name;
        this.intelligence=intelligence;
        this.strength=strength;
        this.health=health;
        
        this.row=INVALID_POS;
        this.col=INVALID_POS;
    }
    
    /**
     * Consctructor de copia
     * @param other Objeto que copiar
     */
    public LabyrinthCharacter (LabyrinthCharacter other){
        this.name=other.name;
        this.intelligence=other.intelligence;
        this.strength=other.strength;
        this.health=other.health;
        
        this.row=other.row;
        this.col=other.col;
    }
    
    /**
     * Comprueba si está muerto el personaje
     * @return Devuelve true si está muerto y false en caso contrario
     */
    public boolean dead(){
        return (this.health<=0);
    }
    
    /**
     * Informa sobre la fila del personaje en el tablero
     * @return Fila de la pos del personaje en el tablero
     */
    public int getRow(){
        return this.row;
    }
    
    /**
     * Informa sobre la columna del personaje en el tablero
     * @return Columna de la pos del personaje en el tablero
     */
    public int getCol(){
        return this.col;
    }
    
    /**
     * Método para consultar el valor de la inteligencia del personaje
     * @return Valor de la inteligencia
     */
    protected float getIntelligence(){
        return this.intelligence;
    }
    
    /**
     * Método para consultar el valor de la fuerza del personaje
     * @return Valor de la fuerza
     */
    protected float getStrength(){
        return this.strength;
    }
    
    /**
     * Método para consultar el valor de la salud del personaje
     * @return Valor de la salud
     */
    protected float getHealth(){
        return this.health;
    }
    
    /**
     * Método para modificar el valor de la salud del personaje
     * @param health Salud nueva a actualizar
     */
    protected void setHealth(float health){
        this.health=health;
    }
    
    /**
     * Definimos la posición del personaje en el tablero
     * @param row Fila en el tablero
     * @param col Columna en el tablero
     */
    public void setPos(int row, int col){
        this.row=row;
        this.col=col;
    }
    
    /**
     * Muestra las características del personaje
     * @return devuelve una cadena con la información del personaje y su posición
     */
    @Override
    public String toString(){
        final String FORMAT = "%.6f";
        String toReturn=this.name+"["+"i:"+ String.format(FORMAT, this.intelligence) + ", s:"+String.format(FORMAT, this.strength);
        toReturn+=", h:"+String.format(FORMAT, this.health)+", p:("+this.row+", "+this.col+")]";
        
        return toReturn;
    }
    
    /**
     * Decrementa en uno la vida del monstruo
     */
    protected void gotWounded(){
        this.health--;
    }
    
    /**
     * Método que indica la fuerza del ataque del personaje
     * @return Devuelve el poder del ataque
     */
    public abstract float attack();
    
    /**
     * Método que se encarga de gestionar la defensa del personaje frente
     * a un ataque.
     * @param attack Intensidad del ataque recibido.
     * @return Devuelve true si el personaje ha muerto y false en caso contrario.
     */
    public abstract boolean defend (float attack);
}

File: Irrgarten_Ruby/labyrinth_character.rb
# encoding: UTF-8


module Irrgarten

    # Clase que representa un personaje del laberinto, ya sea monstruo ({Monster}) o jugador ({Player}).
    #
    # @abstract Implementar los métodos {#attack} y {#defend} en las clases hijas.
    #
    # @author Joaquin Avilés de la Fuente
    # @author Arturo Olivares Martos
    class LabyrinthCharacter

        # Posición inválida
        @@INVALID_POS = -1

        # Constructor de la clase {LabyrinthCharacter}. Inicializa los atributos de la clase.
        # La posición inicial del personaje es inválida
        #
        # @param name [String] Nombre del personaje
        # @param intelligence [float] Inteligencia del personaje
        # @param strength [float] Fuerza del personaje
        # @param health [float] Salud del personaje
        def initialize(name, intelligence, strength, health)
            @name = name.to_s
            @intelligence = intelligence.to_f
            @strength = strength.to_f
            @health = health.to_f

            @row = @@INVALID_POS
            @col = @@INVALID_POS
        end


        protected
        # Consultor de @name
        # @return [String] nombre del personaje
        attr_reader :name

        # Consultor de @intelligence
        # @return [float] inteligencia del personaje
        attr_reader :intelligence

        # Consultor de @strength
        # @return [float] fuerza del personaje
        attr_reader :strength

        # Consultor y modificador de @health
        # @return [float] salud del personaje
        attr_accessor :health

        public
        # Consultor de @row
        # @return [int] fila de la posición del personaje
        attr_reader :row

        # Consultor de @col
        # @return [int] columna de la posición del personaje
        attr_reader :col

        # Modificador de la posición del personaje
        #
        # @param row [int] fila de la posición del personaje
        # @param col [int] columna de la posición del personaje
        def pos(row, col)
            @row = row
            @col = col
        end

        # Método que busca asemejarse a un constructor de copia.
        # Copia los atributos de un personaje a otro.
        #
        # @param other [LabyrinthCharacter] personaje al que se le copiarán los atributos
        # @note Al terminar se copian los atributos del personaje **other** al personaje que llama al método
        def copy(other)
            @name = other.name
            @intelligence = other.intelligence
            @strength = other.strength
            @health = other.health
            pos(other.row, other.col)
        end

        # Método que informa sobre si un personaje ha muerto o no.
        # Un personaje ha muerto si su salud es menor o igual que 0
        #
        # @return [boolean] **true** si el personaje ha muerto, **false** en caso contrario
        def dead
            return @health <= 0
        end

        protected
        # Este método decrementa en una unidad el atributo que representa la salud del personaje.
        def got_wounded
            @health -= 1
        end

        public


        # Método que genera una cadena de caracteres con la información del personaje
        #
        # @return [String] cadena de caracteres con la información del personaje
        def to_s

            # Formato para mostrar los datos flotantes del personaje
            formato='%.6f'

            return "#{@name}[i:#{format(formato,@intelligence)}, s:#{format(formato,@strength)}, "+
            "h:#{format(formato,@health)}, p:(#{@row}, #{@col})]"
        end

        # Método que realiza un ataque.
        #
        # @abstract Método abstracto. Implementar en las clases hijas.
        # @raise [NotImplementedError] Si se llama en esta clase.
        def attack
            raise NotImplementedError
        end

        # Método que defiende al personaje.
        #
        # @abstract Método abstracto. Implementar en las clases hijas.
        # @raise [NotImplementedError] Si se llama en esta clase.
        def defend
            raise NotImplementedError
        end


    end # class
end # module
